### DISCUSSION TITLE
re: recursive type expansion

### DISCUSSION THREAD

**Discussion by refactorized** [ricokahler/sanity-codegen#178] [General]
Hi again, 

I was trying to implement types using the reference expanding utility type you provided yesterday (thanks again) here: 
https://github.com/ricokahler/sanity-codegen/issues/175#issuecomment-900805132

But when I apply it to a type that comes out of the codegen, I just get the same type back. It seems that I have to union the type from the codegen with `Record<string, unknown>` to get the type to resolve correctly. Otherwise the last conditional fails, and I just get the type back. 

``` typescript
type ResolvedSanityReferences<T> =
  // match `SanityKeyedReference` and unwrap via `infer U`
  T extends SanityKeyedReference<infer U>
    ? U
    : // match `SanityReference` and unwrap via `infer U`
    T extends SanityReference<infer U>
    ? U
    : // match arrays, unwrap with `T[number]`,
    // recursively run through `ResolvedSanityReferences`
    // then re-wrap in an another array
    T extends any[]
    ? Array<ResolvedSanityReferences<T[number]>>
    : // match objects, then utilize map types to
    // recursively run children through `ResolvedSanityReferences`
    T extends Record<string, unknown>
    ? {[P in keyof T]: ResolvedSanityReferences<T[P]>} // <-------------------- not matching
    : T;
```

I'm not sure why this doesn't work. As far as I can tell, the generated type should fulfill `Record<string, unknown>` but I am a little unsure about how unknown gets treated in type math. The following is what my code looks like now, and it works, but do you know any way I can use it without unioning the generated type?

``` typescript
import type {TextAndImageBlock} from '@schema/types';
import type {ResolvedSanityReferences} from '@data/types';

type TextAndImageBlockData = ResolvedSanityReferences<
  TextAndImageBlock & Record<string, unknown>
>;

const test = (a: TextAndImageBlockData) => {
  let b = a.desktopImage.asset.assetId;  // <----------- asset is properly de-referenced here
};

export default TextAndImageBlockData;
```

**Comment by ricokahler** [comment]
could you provide the generated types along with the use case that doesn't work? I can better help debug then

**Reply by refactorized** [reply]
Here is the main type, `textAndImageBlock`, straight from the codegen with comments removed. I also include the definitions for `Link` and `BasicText`:
``` typescript
export interface TextAndImageBlock extends SanityDocument {
  _type: "textAndImageBlock";

  blockType?: string;
  header?: string;
  subHeader?: string;
  body?: BasicText;
  desktopImage?: {
    _type: "image";
    asset: SanityReference<SanityImageAsset>;
    crop?: SanityImageCrop;
    hotspot?: SanityImageHotspot;
  };
  mobileImage?: {
    _type: "image";
    asset: SanityReference<SanityImageAsset>;
    crop?: SanityImageCrop;
    hotspot?: SanityImageHotspot;
  };
  buttonText?: string;
  buttonLink?: Link;
}

export type BasicText = Array<SanityKeyed<SanityBlock>>;

export type Link = {
  _type: "link";
  blockType?: string;
  refLink?: SanityReference<Page>;
};
```

Possibly there is a recursion problem with Link referencing Page which can contain a TextAndImageBlock.  I am going to try to change that to just use slugs and see if that helps at all.  Everything else is types from sanity-codegen.

**Update:** removing the Page reference from Link didn't seem to help new Link type is:

``` typescript
export type Link = {
  _type: "link";
  blockType?: string;
  slug?: { _type: "slug"; current: string };
  url?: string;
};
```

**Reply by ricokahler** [reply]
so i tried always using the mapped type (instead of matching conditionally on Record<string, unknown>` and i think that works

[see here](https://www.typescriptlang.org/play?#code/PQKhCgAIUglBTADgJ3gZ3gOwC5sgQ0lQDN5VMBjeSAS00gGV9MbsBPSbAewMy+wAWZSFmys2AOkgA5ftUH5snIVBgBzLGRoVObRNRp402ZNuwAbDsS7JIAFT3wGFU4iUBbeNkIoua5Pju7nRqEqrA4MDAMgDydgCiAFzK1AD6drR41rZcmJaQ2br6aC40bpCe3pHR6OZ02AC0ACaG+ABG5vANmPAAHo11PZAAAuzFpW4NtfXAfA0ArpjzGE0NAG74yGjgfYg2SmPUTCzsCKTkVAA86QB8kAC8kADeUJBvqYfJAOQkZFhUXwA3K93iRksZTJg1MCAL7AyJgVRwJCoDA4PCEX4XAz0Y7iTg8Zj8IS2UT4gDurAEBEgAGt4JIZHJlIoUkiND1TDpDplIBCzPlCg59M5XB4vD5kH4AkEQmFoBEorEEslBGkMoYCjZILlBdrDiUxRUJdURGhBo0Wmh2p1un0BnRqKNHIayo1pjhZlwFksVutNttdvsikdmOIANIM+BNM5-Sjwa52O6PF5vd6fSA-eDnf7wIEgyCpelscEmELAtOFsF8stQ2HwoPIJT1MjEfBURhh9gAES4FHmnhwzwLqRoTVLkOhI5c8EU0YAgtgJ+WR6g1su6yP5ogmnOmouN1OYeAT43mzhW+3Qyc2ABJdz4DQAWQlu+8w8rH0c32tN4kNAfZ9X0UfB80rFpBzQGhcjQZI8XYe9H3gbsAKwKCYIrNMBHwNB53MRBsOSNouC4TpmEwt5DBiRB8AAR3meAiJIsjMAoyBzFospDzYmjOmwbBGM7G9EI0AAFfA+IE2FT16PYm1oC9kDbDt4LvQDkNQzB0K0j80y-fQfy7SQAKQlDIOgrSwLTHD9AobBYEUaDkiWdw2jINihBoNQBCXSAXLc5A2MpJpBGcgcAuknZZODFslKvITxBE+BxMk6hUz0jMvl-cR-3UlKvAEqy3l3ZBaSfeYBPHBKELyiSCqcclFAoAQ2JK2kADUaDaAIcDgoykvy-iGqalqCyaLhgkwZhfNUga6qGhhGuwZq2LqbzsHKyq+uE2rUsWkbVq8nzOu66btsS3b6v25bRsrdwKujc6aqQwaBOulaCzWLqepm-rLoWpaPuPKK5PPAS4pUv6Xvmt7AepdL02-TNsvYXLob2uGisgNp21pfwuEWKr+U3StsngfHCe4gs9kQeZzEc3Iwtc9yCzECxBOJo8TxBmLFOU68LqQ+c0AwJR7SwJo8FU3t+0HJQEcLTKUeM9ThdFrGcNF28idrKdK3FrSLKpu7gO8J61KQl9vDffA2OCTxhQ53W2JsLy6AkgAxGhOimzxjbTGjQprSc2LQbCAEZsLD-23iggAvQT-JZytt3MLh8CabWY8geZkHMKngbPBTwf56qLY0ABhKVEF0xGDORoy0cr6useI-iJqZiKC06YhfKTwKC1MdbO+TtNuEQEeB8L6L5Ni0vZvUgAJfg0D2eWRyVxuTI0ZfcDXrHPOHvzwtHt5gqD-u2N6Se2JLY-mank8lQ6PtaQxVAim0CT8mwzAmk6Jo2MOCEBfhQWkkAKC5DWGQbS7EuqmgABTwAkKEbGacwGQAADoAANhjK2AKA2kDRIEXhwA0bgDRUDtmwDggAlHyHg5JqC6g4DyNaPlLA81nnzeKqkABC6DwEK30oJL4hCsYAG1iyHgALrJGYGwSKPIEBoFItApoqlYzYjQJcJMDwCxKgfDdSA2DVKRjYNGLRuZsG8EAYsckAQa5fUINgug5xIAAFVsEFgyAbSWZdzGWOzHGK4bjhAeJuAWAA-J4k8lZDEjRMZo4J2IbHMDsZgBx+AnE0BcWE2wXiCzJF8f0CWUsjJWPjJcfJnjImVhiR4uJaYEnGM2AENgaAAA0OdMmOMgJSQQJi7ASP7jI7BnSDHRFQP2LYNBoH5GQIsZQUp5jeRMSotR0Zkk5njGgbx8TohqnoKgBoWSa50F4LwYkwg2n4DYEU+wIhSl-wxJgNgEiZHRMgPOZA7TLgbPMOo7ZIT0C6JGSfZAMibiRMmQUd25h8gVW9jQBOFRsn6EAQaAksLpm5ygvMjgiz6DNW9k0cgyyCZrOwQCoFFSUm5j2Q8p4kAJGiQUnSBkXBiD2DkciVRgKtl0p2VQHRwzRJQsgNPUGxdLwdjsPaecf8koCNfk8gSLyy4ywHKIWuiskYACIBL9EVZndSKqwH6vhJWQhjsonZyEBnMgdrg4rkrGgeYbRF6ziaE67OxEmhsGdXwnC2h5X9FaugWk48krOoVp+DM+rt7wEtQWSsmsvDm0qVcBeQsRZeBuGxSsLguCIGdTm5uJbC1YRXmvMtUMd41v4GxOEBZ3BcDaN7eAMbkhxoygapNKbKxprzb9G8WaEzlvgGrfNVa3jFtLebJKVdK2purXvfgdadpIV3qvJtBYW3WoqtwTAYbsDOs5mxNoR7cgABk6C0mdXezAtJIpFx5GYqMTRdHJkeX4vA7aABW8A7KQBiRkAAZM8Qs0iXVQklZAYpDYZ4HEcJAYNUEKCnoeN835dzLgfosV+-hgjoXwgSWwNy-ScIVEMFBKEUTwA8nEhobDTwRHfBohoL4Urgw8ifeAlMG8DWDFpIOtMNrHDnudoPbM-HN3iHHZcZj8AC3gBbYx1Dp6TXKsEd2EC2GaWCrHfS3ZuiFVKrNSR+EPqKD0w-iQ4wkBr72HM6apC5raR6e8PCXoEgfVoCjSWmNEh03YAkIHAQQA)

here's the updated type

```ts
type ResolvedSanityReferences<T> =
    // match `SanityKeyedReference` and unwrap via `infer U`
    T extends SanityKeyedReference<infer U>
    ? U

    // match `SanityReference` and unwrap via `infer U`
    : T extends SanityReference<infer U>
    ? U

    // match arrays, unwrap with `T[number]`,
    // recursively run through `ResolvedSanityReferences`
    // then re-wrap in an another array
    : T extends any[]
    ? Array<ResolvedSanityReferences<T[number]>>

    // finally utilize mapped types to
    // recursively run children through `ResolvedSanityReferences`
    : { [P in keyof T]: ResolvedSanityReferences<T[P]> }
```

**Reply by refactorized** [reply]
Yep! That works.  Makes sense since my union was making that last clause tautological anyways. Thanks a bunch!

**Reply by refactorized** [reply]
Future iteration of this stack will likely involve more verbose groq queries which explicitly expand references using select and project logic, and lean on 1.0+ versions of sanity-codegen with its query based generation. I will be watching this repo closely.

**Reply by nahtnam** [reply]
Hello! Sorry to bring up an old thread, but I noticed that this breaks when you have an array of references and inside that you have an array of references. I tried to play around with the type but I couldn't figure it out

[Here is a reproduction](https://www.typescriptlang.org/play?#code/PQKhCgAIUglBTADgJ3gZ3gOwC5sgQ0lQDN5VMBjeSAS00gGV9MbsBPSbAewMy+wAWZSFmys2AOkgA5ftUH5snIVBgBzLGRoVObRNRp402ZNuwAbDsS7JIAFT3wGFU4iUBbeNkIoua5Pju7nRqEqrA4MDAMgDydgCiAFzK1AD6drR41rZcmJaQ2br6aC40bpCe3pHR6OZ02AC0ACaG+ABG5vANmPAAHo11PZAAAuzFpW4NtfXAfA0ArpjzGE0NAG74yGjgfYg2SmPUTCzsCKTkVAA86QB8kAC8kADeUJBvqYfJAOQkZFhUXwA3K93iRksZTJg1MCAL7AyJgVRwJCoDA4PCEX4XAz0Y7iTg8Zj8IS2UT4gDurAEBEgAGt4JIZHJlIoUkiND1TDpDplIBCzPlCg59M5XB4vD5kH4AkEQmFoBEorEEslBGkMoYCjZILlBdrDiUxRUJdURGhBo0Wmh2p1un0BnRqKNHIayo1pjhZlwFksVutNttdvsikdmOIANIM+BNM5-Sjwa52O6PF5vd6fSA-eDnf7wIEgyCpelscEmELAtOFsF8stQ2HwoPIJT1MjEfBURhh9gAES4FHmnhwzwLqRoTVLkOhI5c8EU0YAgtgJ+WR6g1su6yP5ogmnOmouN1OYeAT43mzhW+3Qyc2ABJdz4DQAWQlu+8w8rH0c32tN4kNAfZ9X0UfB80rFpBzQGhcjQZI8XYe9H3gbsAKwKCYIrNMBHwNB53MRBsOSNouC4TpmEwt5DBiRB8AAR3meAiJIsjMAoyBzFospDzYmjOmwbBGM7G9EI0AAFfA+IE2FT16PYm1oC9kDbDt4LvQDkNQzB0K0j80y-fQfy7SQAKQlDIOgrSwLTHD9AobBYEUaDkiWdw2jINihBoNQBCXSAXLc5A2MpJpBGcgcAuknZZODFslKvITxBE+BxMk6hUz0jMvl-cR-3UlKvAEqy3l3ZBaSfeYBPHBKELyiSCqcclFAoAQ2JK2kADUaDaAIcDgoykvy-iGqalqCyaLhgkwZhfNUga6qGhhGuwZq2LqbzsHKyq+uE2rUsWkbVq8nzOu66btsS3b6v25bRsrdwKujc6aqQwaBOulaCzWLqepm-rLoWpaPuPKK5PPAS4pUv6Xvmt7AepdL02-TNsvYXLob2uGisgNp21pfwuEWKr+U3StsngfHCe4gs9kQeZzEc3Iwtc9yCzECxBOJo8TxBmLFOU68LqQ+c0AwJR7SwJo8FU3t+0HJQEcLTKUeM9ThdFrGcNF28idrKdK3FrSLKpu7gO8J61KQl9vDffA2OCTxhQ53W2JsLy6AkgAxGhOimzxjbTGjQprSc2LQbCAEZsLD-23iggAvQT-JZytt3MLh8CabWY8geZkHMKngbPBTwf5yAACEpXpegDcl6q2BlgdRF0yj1NgyB52QAI2EuVTIzYaNY2xHuoY0NWvBuG5wEL6L5Ni0vZvUgBhKVEGbxWkayoy0Y0ZeuEQLHiP4iamYigtOmIXyk8CgtTHWk-k7TbhEHv6-p9B4vL0hnakIACX4NA9jyxHErLeJkNB-1wIArGnk75+XCg-N4wUg5XzYr0F+bESxwOZq-E8SoOh9lpBiVARRtASXyNhTATROhNGxhwQg+CKC0kgBQXIawyDaXYl1U0AAKeAEhQjYzTowyAAAdAABsMZWwAGG0gaCwi8OAGjcAaKgds2BxEAEo+Q8HJNQXUHAeRrR8pYHms8+bxVUmXIRTCFb6UEl8GRWMADaxZDwAF1kjMDYJFHkCA0CkTYU0VSg9cxoEuEmB4BYlQPhupAMRvcowxmzHGKgYjeA0MWOSAIq8vqEDEXQc4kAACqYiCwZBrlLIyfcB7JKHgU4QRTJ6VgAPzFJPJWaJI04nBNqbmNJzAMmYCyfgHJNA8n1NsCUgsyRyn9AlpUm8IT4yXAmcUppaZWlFPaWmTpsTNhdzQAAGhzkM7JkBKSCDiXYJxV83FiMOVE6IqB+xbBoGw-IyBFjKClPMbycS-EBOjD0nM8Y0ClI6dENU9BUANGGavOgvBeDEmEPs-AbBpn2BEHMyhGJMBsCcW4gsrSO5d0uAC8wgTgUpPQOEm58DkBuIntst4SpiDu3MPkCq3saAJwqCM-QNCDQEkeUQeALyoLvI4J8+gzVvZNHIN8gmfyxHkspUZJZVAwUYqeJAJxokFJ0gZFwYg9gPHIn8RSoF6remgtpaJRlkA368xLvFOw9p5yUKSlYghWKBI4rrg3OWa87HJAAEQCX6B6zO6lvWMNDfCSsMjHbNOzkIDOZAU3BxXJWNA8w2g-1nE0DN2diJNDYJmsuOFtBuv6K1dAtIn5JUzQrT8GZQ1gPgPGgslZNZeHNhqhMC8hYi3HmxSsLg96ZqHTvFeY6sL-0AVOke8AIEAP4GxOEBZ3BcDaN7eATbkgtoykjdt6ku2Vh7SO36iybVXGnfAMe2AbhzreBOxAS7v4zr3i+yAAgF38A-YLcB-7sAbrYm0Cq3BMA1uwJmzm4HIO5AADJ0FpJmlDmBaQJrTN1LgVdkgkrRcPG81SkkgquBXPDWAJ6RSLjyBJ-cmjhOTJiipOo2gACsxVKFaRkAAZM8Qsris1QkdZAGZDYZ4HEcOXKtFAYMPHbp3IjDHozEfELG2kNHcHRAfGwNy5ycIVEMFBKEzTwA8nEhoRTTwQ2ZhohoL4Tr5I8gw0wlMwCT2DFpOenD1jk3ZxIG5wDpxb0Jis-AZ9U94Q8hg1Gr11juwgUU6qq1N7yM0ri56mN1iovgCLRQemxD5HGEgGg+w7rstIU00l7w8JegSFw1XJxAAGNx290Ctfa7ncwQA). Keep an eye on the `broken` and `Broken` values

Do either of you have any idea?

cc @ricokahler

**Reply by nahtnam** [reply]
This is the solution i came up with:

```
type ResolvedSanityReferences<T> =
    // match `SanityKeyedReference` and unwrap via `infer U`
    T extends SanityKeyedReference<infer U>
    ? ResolvedSanityReferences<U>

    // match `SanityReference` and unwrap via `infer U`
    : T extends SanityReference<infer U>
    ? ResolvedSanityReferences<U>

    // match arrays, unwrap with `T[number]`,
    // recursively run through `ResolvedSanityReferences`
    // then re-wrap in an another array
    : T extends any[]
    ? Array<ResolvedSanityReferences<T[number]>>

    // finally utilize mapped types to
    // recursively run children through `ResolvedSanityReferences`
    : { [P in keyof T]: ResolvedSanityReferences<T[P]> }
```

Simply added `ResolvedSanityReferences` around the first two cases as well. That makes it recursive on the inner references as well
